zkevm-circuits/src/evm_circuit.rs

```diff
@@ -944,12 +1049,93 @@ mod test {
         circuit::{Layouter, SimpleFloorPlanner},
         plonk::{Circuit, ConstraintSystem, Error},
     };
+    extern crate hex;
+    extern crate num;
+    use bus_mapping::evm::OpcodeId;
 
     #[derive(Clone)]
     pub(crate) struct TestCircuitConfig<F> {
         evm_circuit: EvmCircuit<F>,
     }
 
+    // contruct byte codes from compiled by contract,  i.e. bytecode_source = "6002565b00";
+    pub(crate) fn assgin_byte_table(bytecode_source: &str) -> Vec<[u32; 5]> {
+        let byte_codes = hex::decode(bytecode_source).expect("Decoding failed");
+        // TODO: add keccak hash (byte_codes)
+        let code_hash = 0 as u32;
+        let mut i = 0;
+        let mut push_x = 0;
+        let mut bytecode_table = Vec::<[u32; 5]>::new();
+        for byte in byte_codes.iter() {
+            if push_x != 0 {
+                let push_rindex = push_x;
+                for x in 0..push_rindex {
```

Comment:
Not important because this is just test code, but this could go out or range because without code validation (https://eips.ethereum.org/EIPS/eip-3670, which isn't used yet) the push data is not guaranteed to be there.

Reply:
here provides two helper methods to construct  valid byte code table elements,  one is assgin_byte_table which load hex byte code sequence from compiled contract bytecode output, i.e. '608060405234801....', the other is assgin_byte_table_step, which constructs table from evm circuit ExecutionSteps , means collect bytes from current gadget test, so that each test can mock its own byte code table independently, this way can be  avoid changing existing tests to align with one global table.

Reply:
```
for x in 0..push_x {
    bytecode_table.push([code_hash, i + x, 0, *byte as u32]);
}
```
This pushes the same byte `push_x` times because `byte` remains unchanged in this loop if I understand correctly, which I don't think is the correct behavior as each push would need to push the following byte. Without EIP-3670 there is also no guarantee that byte is in the bytecode, so a separate loop for this kind of complicates things as well I feel. I think it's easier to implement this using a simple single loop on the bytes to not run into any of these issues (see my bytecode unrolling PR as an example).

Reply:
right , this method is obsolete now as currently 'TestCircuit' loads bytes from ExecutionSteps within each gadget 's unit test. I am going to remove this one, will add back until it's real needed in the future.